File: utils/docker.go
package utils

import (
	"os"
	"os/exec"
)

// RunOptions is options for DockerRun
type RunOptions struct {
	Name   string
	Port   string
	Image  string
	Env    []string
	Volume string
}

// DockerRun starts a container
func DockerRun(options RunOptions) error {
	args := []string{
		"run",
		"--name",
		options.Name,
		"-p",
		options.Port + ":" + options.Port,
		"--rm",
		"-d",
		"-v",
		options.Volume,
	}
	for _, v := range options.Env {
		args = append(args, "-e", v)
	}
	args = append(args, options.Image)

	cmd := exec.Command(
		"docker",
		args...,
	)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin
	return cmd.Run()
}

// DockerStop stops container by name
func DockerStop(name string) error {
	cmd := exec.Command("docker", "stop", name)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin
	return cmd.Run()
}

// DockerExec is basically a shorthand of `docker exec`
func DockerExec(name string) error {
	cmd := exec.Command("docker", "exec", "-it", name, "/bin/bash")
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin
	return cmd.Run()
}

File: src/docker.ts
import { spawnSync, execSync, ExecSyncOptions } from "child_process"
import { printPort } from "./utils"

let requireSudo: boolean | undefined

export function checkDockerPermission() {
  if (typeof requireSudo === "boolean") return requireSudo

  const cmd = spawnSync("docker", ["ps"])
  const output = cmd.stderr.toString()

  if (cmd.status === 0) {
    requireSudo = false
  } else {
    if (output.includes("permission denied")) {
      requireSudo = true
    } else {
      throw new Error(output)
    }
  }

  return requireSudo
}

export function runDockerCommand(
  command: string,
  { stdio }: { stdio?: "inherit" | "pipe" } = {},
) {
  const requireSudo = checkDockerPermission()
  const options: ExecSyncOptions = {
    stdio: stdio || "inherit",
  }
  if (requireSudo) {
    return execSync(`sudo -- su -c '${command}'`, options)
  }
  return execSync(command, options)
}

export function dockerRun({
  serviceName,
  image,
  volumes,
  port,
  env,
  extraArgs,
}: {
  serviceName: string
  image: string
  volumes?: string[]
  port?: Array<`${number}:${number}`>
  env?: string[]
  extraArgs?: string[]
}) {
  const args: string[] = ["run", "--name", `doko_${serviceName}`, "--rm", "-d"]

  if (port) {
    port.forEach((p) => {
      args.push("-p", p)
    })
  }

  if (volumes) {
    for (const vol of volumes) {
      args.push(`-v`, vol)
    }
  }

  if (env) {
    for (const value of env) {
      args.push("-e", value)
    }
  }

  args.push(image)

  if (extraArgs) {
    args.push(...extraArgs)
  }

  const command = `docker ${args.join(" ")}`

  runDockerCommand(command)

  if (port) {
    printPort(port)
  }
}

export function dockerStop(serviceName: string) {
  runDockerCommand(`docker stop doko_${serviceName}`)
}

export function getRunningDockerContainers() {
  const output = runDockerCommand(`docker ps --format "{{json .}}\\n\\n"`, {
    stdio: "pipe",
  })
    .toString()
    .trim()

  if (!output) return []

  const services: {
    Names: string
    Image: string
    Ports: string
    Size: string
    RunningFor: string
  }[] = output.split(/\n{2,}/).map((v) => JSON.parse(v))

  return services
}

export function isDockerContainerRunning(containerName: string) {
  const running =
    runDockerCommand(
      `docker ps --filter "name=${containerName}" --format "{{json .}}"`,
      { stdio: "pipe" },
    )
      .toString()
      .trim() !== ""

  return running
}

export function dockerExec(serviceName: string) {
  runDockerCommand(`docker exec -it doko_${serviceName} /bin/bash`)
}
